### PR TITLE
Admin ghosts can now interact with stuff

### PR DESCRIPTION
**Pull request by SweptWasTaken**
## About the PR 

I don't know if this is the right way to do this. I also don't know if explicitly adding HandComponent to the admin observer is the way to go.

**Changelog**

:cl:
- add: Admin ghosts can interact with stuff

### CODE REVIEW

File: Content.Server/Ghost/Components/GhostComponent.cs [hunk 1]
```diff
@@ -38,6 +43,15 @@ public bool CanReturnToBody
                 Dirty();
             }
         }
+        public bool CanGhostInteract
+        {
+            get => _canGhostInteract;
+            set
+            {
+                _canGhostInteract = value;
+                Dirty();
```
Comment (Zumorica) [inline Content.Server/Ghost/Components/GhostComponent.cs:52]: You're calling Dirty here but you're not adding this variable to the ghost state.
Our system sadly cannot rely on magic, we need to be explicit about the variables we sync in states

**Review by Zumorica — CHANGES_REQUESTED** [review]
This is probably gonna conflict with https://github.com/space-wizards/space-station-14/pull/4159

File: Content.Server/Ghost/Components/GhostComponent.cs (hunk 1 above)
Comment (metalgearsloth) [reply Content.Server/Ghost/Components/GhostComponent.cs:52]: Should also guard it with if (_canGhostInteract == value) return;

**Comment by SweptWasTaken** [comment]
> 
> 
> This is probably gonna conflict with #4159

Waiting for that to be merged and then i'll address.

File: Content.Server/Ghost/Components/GhostComponent.cs (hunk 1 above)
Comment (metalgearsloth) [reply Content.Server/Ghost/Components/GhostComponent.cs:52]: Also SharedGhostComponent has this property; you should override it instead.

**Comment by metalgearsloth** [comment]
Currently you don't have any hands on adminobserver so can't actually pick stuff up but the interaction part works. I've pinged smug on handling it (as it means adminobservers would require body + bodypart comp)

**Comment by metalgearsloth** [comment]
Talked to Smug and we'll just add the requisite body parts for now and realise it's shitcode later.

File: Content.Shared/Ghost/SharedGhostComponent.cs [hunk 2]
```diff
@@ -52,18 +79,21 @@ public override void HandleComponentState(ComponentState? curState, ComponentSta
...
 
         public HashSet<string>? LocationWarps { get; }
 
         public Dictionary<EntityUid, string>? PlayerWarps { get; }
 
         public GhostComponentState(
             bool canReturnToBody,
+            bool canGhostInteract,
             HashSet<string>? locationWarps = null,
             Dictionary<EntityUid, string>? playerWarps = null)
```
Comment (metalgearsloth) [inline Content.Shared/Ghost/SharedGhostComponent.cs:92]: These need removing as they were jiggled around with the refactor (this is a reminder for futuresloth).

**Comment by SweptWasTaken** [comment]
R u coming back to this?

**Comment by metalgearsloth** [comment]
> 
> 
> R u coming back to this?

image

yes

**Comment by metalgearsloth** [comment]
This is finally ready for another soul to review.

File: Content.Shared/Body/Components/SharedBodyComponent.cs [hunk 3]
```diff
@@ -474,51 +472,8 @@ public IEnumerable<(SharedBodyPartComponent part, IBodyPartProperty property)> G
             }
         }
 
-        private void CalculateSpeed()
```
Comment (mirrorcult) [inline Content.Shared/Body/Components/SharedBodyComponent.cs:477]: why has this just been removed

Comment (metalgearsloth) [reply Content.Shared/Body/Components/SharedBodyComponent.cs:477]: image

https://discord.com/channels/310555209753690112/310555209753690112/877189830558253077

**Pull request merged**